### PR TITLE
ws: Translate login page

### PR DESCRIPTION
**Pull request by mvollmer**

### CODE REVIEW

**Comment by mvollmer** [comment]
This embeds the po data right into login.<lang>.html.

I guess we want to keep the po data in their own files for easier independent replacement/adding, right?
I'll make it so that `send_login_html` embeds it at run-time.

**Comment by mvollmer** [comment]
We should do something about fuzzy translations, they are pretty wrong most of the time.

**Comment by mvollmer** [comment]
> I'll make it so that send_login_html embeds it at run-time.

Done with the FIXUP commit.  I like that better than the original approach, but I'll leave both visible for review.

File: src/ws/login.js [hunk 1]
```diff
@@ -14,6 +14,26 @@ var phantom_checkpoint = phantom_checkpoint || function () { };
             oauth.ErrorParam = "error_description";
     }
 
+    function gettext(key) {
+        console.log(key);
```
Comment (mvollmer) [inline src/ws/login.js]: Debug noise.

File: src/common/cockpitwebresponse.c [hunk 2]
```diff
@@ -1572,7 +1572,7 @@ cockpit_web_response_negotiation (const gchar *path,
 
   while (!bytes)
     {
-      if (locale && shorter)
+      if (locale && shorter && g_strcmp0 (locale, shorter))
```
Comment (petervo) [inline src/common/cockpitwebresponse.c]: Should this be ```g_strcmp0 (locale, shorter) != 0``` or is it supposed to be checking equal? ```== 0```

Comment (mvollmer) [inline src/common/cockpitwebresponse.c]: It should be `g_strcmp0 (locale, shorter) != 0`, I'll clarify that.

**Comment by mvollmer** [comment]
> This seems to be missing an integration test?

Yep.

File: src/ws/login.js (hunk 1 above)
Comment (mvollmer) [inline src/ws/login.js]: Gone.

**Comment by mvollmer** [comment]
> This seems to be missing an integration test?

Added a simple smoke test.

**Comment by mvollmer** [comment]
`format` in login.js does not work the way I thought it works.

**Comment by mvollmer** [comment]
Firefox complains because the inserted po data move the <meta http-equiv> to far down in the file: 

```
The page was reloaded, because the character encoding declaration of the HTML document was not found when prescanning the first 1024 bytes of the file. The encoding declaration needs to be moved to be within the first 1024 bytes of the file.  f25:9090:144
```

**Comment by mvollmer** [comment]
> Firefox complains because the inserted po data move the <meta http-equiv> to far down in the file:

Fixed.

**Comment by mvollmer** [comment]
Tests need updating.

**Comment by mvollmer** [comment]
> Tests need updating.

This was interesting.  What happended was that when the language was "en", the negotiation logic would try "login.po.en.html" and  "login.po.html", as expected.  But "login.po.html" didn't exist (which is the bug), and it would then go on and try "login.html".  Thus, it would insert another copy of "login.html" into itself, which would cause two #option-group elements to exist, which would break the tests.

@stefwalter, I didn't expect the negotiation logic to remove extensions, such as serving "login.html" when asked for "login.po.html".  Is that intentional?

**Comment by dperpeet** [comment]
we want this to go into the next release

File: src/common/cockpitwebresponse.c [hunk 3]
```diff
@@ -1572,7 +1572,7 @@ cockpit_web_response_negotiation (const gchar *path,
 
   while (!bytes)
     {
-      if (locale && shorter)
+      if (locale && shorter && g_strcmp0 (locale, shorter) != 0)
```
Comment (dperpeet) [inline src/common/cockpitwebresponse.c]: as discussed on irc, this needs a comment (and maybe an updated commit message)

@mvollmer's suggestion: `only consider the shorter version of a locale when it is different from the full string`

I'm also not entirely sure about the logic of what is checked first.

File: src/ws/login.js [hunk 4]
```diff
@@ -14,6 +14,31 @@ var phantom_checkpoint = phantom_checkpoint || function () { };
             oauth.ErrorParam = "error_description";
     }
 
+    var fmt_re = /\$\{([^}]+)\}|\$([a-zA-Z0-9_]+)/g;
```
Comment (dperpeet) [inline src/ws/login.js]: nitpick: this is removed in one commit and added back in later...

**Comment by dperpeet** [comment]
As discussed on irc with @mvollmer, full locale should have precedence over short version.

**Comment by dperpeet** [comment]
Thanks for doing this and adding the comments @mvollmer !

**Comment by dperpeet** [comment]
video: https://youtu.be/_4jQ6NYSL-c

**Pull request closed without merging**